File: mobile/ios/TrainBlink/TrainBlink/Models/Discovery.swift
//
//  Discovery.swift
//  TrainBlink
//
//  BLE discovery and nearby traveler models
//

import Foundation

/// Distance estimation based on RSSI signal strength
enum DistanceEstimate: String, Codable {
    case veryClose = "Very Close (0-2m)"
    case close = "Close (2-10m)"
    case medium = "Medium (10-30m)"
    case far = "Far (30-50m)"
    case veryFar = "Very Far (50-100m)"

    /// Initialize from RSSI value
    init(rssi: Int) {
        switch rssi {
        case -50...0:
            self = .veryClose
        case -70...(-51):
            self = .close
        case -85...(-71):
            self = .medium
        case -95...(-86):
            self = .far
        default:
            self = .veryFar
        }
    }

    /// Get emoji icon for distance
    var icon: String {
        switch self {
        case .veryClose: return "👤"
        case .close: return "🚶"
        case .medium: return "🚶‍♂️"
        case .far: return "🏃"
        case .veryFar: return "🏃‍♂️"
        }
    }

    /// Get color for distance indicator
    var color: String {
        switch self {
        case .veryClose: return "green"
        case .close: return "lightGreen"
        case .medium: return "yellow"
        case .far: return "orange"
        case .veryFar: return "red"
        }
    }

    /// Get short description
    var shortDescription: String {
        switch self {
        case .veryClose: return "Very Close"
        case .close: return "Close"
        case .medium: return "Medium"
        case .far: return "Far"
        case .veryFar: return "Very Far"
        }
    }
}

/// Discovered user via BLE
struct DiscoveredUser: Identifiable {
    let id: String // BLE Anonymous ID (TB_xxxxx)
    let rssi: Int
    let distance: DistanceEstimate
    let discoveredAt: Date
    var lastSeen: Date

    /// Check if discovery is fresh (within last 30 seconds)
    var isFresh: Bool {
        return Date().timeIntervalSince(lastSeen) < 30
    }

    /// Time since last seen
    var timeSinceLastSeen: TimeInterval {
        return Date().timeIntervalSince(lastSeen)
    }
}

/// Request payload for logging a discovery event
struct LogDiscoveryRequest: Codable {
    let tripRoute: String
    let discoveredUserAnonymousId: String
    let distanceEstimate: String
    let rssi: Int
    let discovererAgeRange: String?
    let discoveredAgeRange: String?
    let discovererGender: String?
    let discoveredGender: String?

    enum CodingKeys: String, CodingKey {
        case tripRoute = "trip_route"
        case discoveredUserAnonymousId = "discovered_user_anonymous_id"
        case distanceEstimate = "distance_estimate"
        case rssi
        case discovererAgeRange = "discoverer_age_range"
        case discoveredAgeRange = "discovered_age_range"
        case discovererGender = "discoverer_gender"
        case discoveredGender = "discovered_gender"
    }
}

/// Discovery analytics response
struct DiscoveryStatsResponse: Codable {
    let status: String
    let data: DiscoveryStats?

    struct DiscoveryStats: Codable {
        let totalDiscoveries: Int
        let uniqueUsers: Int
        let averageRssi: Double
        let mostCommonDistance: String

        enum CodingKeys: String, CodingKey {
            case totalDiscoveries = "total_discoveries"
            case uniqueUsers = "unique_users"
            case averageRssi = "average_rssi"
            case mostCommonDistance = "most_common_distance"
        }
    }
}

File: mobile/android/TrainBlink/app/src/main/java/org/trainblink/app/models/Discovery.kt
package org.trainblink.app.models

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.Date

/**
 * Distance estimation based on RSSI signal strength
 */
enum class DistanceEstimate(val rawValue: String) {
    VERY_CLOSE("Very Close (0-2m)"),
    CLOSE("Close (2-10m)"),
    MEDIUM("Medium (10-30m)"),
    FAR("Far (30-50m)"),
    VERY_FAR("Very Far (50-100m)");

    companion object {
        /**
         * Initialize from RSSI value
         */
        fun fromRssi(rssi: Int): DistanceEstimate {
            return when {
                rssi >= -50 -> VERY_CLOSE
                rssi >= -70 -> CLOSE
                rssi >= -85 -> MEDIUM
                rssi >= -95 -> FAR
                else -> VERY_FAR
            }
        }
    }

    /**
     * Get emoji icon for distance
     */
    val icon: String
        get() = when (this) {
            VERY_CLOSE -> "👤"
            CLOSE -> "🚶"
            MEDIUM -> "🚶‍♂️"
            FAR -> "🏃"
            VERY_FAR -> "🏃‍♂️"
        }

    /**
     * Get color for distance indicator (Material colors)
     */
    val colorHex: String
        get() = when (this) {
            VERY_CLOSE -> "#4CAF50"  // Green
            CLOSE -> "#8BC34A"       // Light Green
            MEDIUM -> "#FFEB3B"      // Yellow
            FAR -> "#FF9800"         // Orange
            VERY_FAR -> "#F44336"    // Red
        }

    /**
     * Get short description
     */
    val shortDescription: String
        get() = when (this) {
            VERY_CLOSE -> "Very Close"
            CLOSE -> "Close"
            MEDIUM -> "Medium"
            FAR -> "Far"
            VERY_FAR -> "Very Far"
        }
}

/**
 * Discovered user via BLE
 */
@Parcelize
data class DiscoveredUser(
    val id: String, // BLE Anonymous ID (TB_xxxxx)
    var rssi: Int,
    var distance: DistanceEstimate,
    val discoveredAt: Date,
    var lastSeen: Date
) : Parcelable {

    /**
     * Check if discovery is fresh (within last 30 seconds)
     */
    val isFresh: Boolean
        get() = (Date().time - lastSeen.time) < 30_000

    /**
     * Time since last seen in milliseconds
     */
    val timeSinceLastSeen: Long
        get() = Date().time - lastSeen.time

    /**
     * Update with new RSSI reading
     */
    fun update(newRssi: Int) {
        rssi = newRssi
        distance = DistanceEstimate.fromRssi(newRssi)
        lastSeen = Date()
    }
}

/**
 * Request payload for logging a discovery event
 */
data class LogDiscoveryRequest(
    @SerializedName("trip_route")
    val tripRoute: String,
    @SerializedName("discovered_user_anonymous_id")
    val discoveredUserAnonymousId: String,
    @SerializedName("distance_estimate")
    val distanceEstimate: String,
    val rssi: Int,
    @SerializedName("discoverer_age_range")
    val discovererAgeRange: String? = null,
    @SerializedName("discovered_age_range")
    val discoveredAgeRange: String? = null,
    @SerializedName("discoverer_gender")
    val discovererGender: String? = null,
    @SerializedName("discovered_gender")
    val discoveredGender: String? = null
)

/**
 * Discovery analytics response
 */
data class DiscoveryStatsResponse(
    val status: String,
    val data: DiscoveryStats?
) {
    data class DiscoveryStats(
        @SerializedName("total_discoveries")
        val totalDiscoveries: Int,
        @SerializedName("unique_users")
        val uniqueUsers: Int,
        @SerializedName("average_rssi")
        val averageRssi: Double,
        @SerializedName("most_common_distance")
        val mostCommonDistance: String
    )
}
